### DISCUSSION TITLE
The needsCustomLayoutForChildren check is missing from the new React Native architecture

### DISCUSSION THREAD

**Discussion by grahammendick** [reactwg/react-native-new-architecture#52] [Q&A] [answered]
I've started updating my navigation library to work with the new architecture and I've run into a problem on Android. The new React Native architecture doesn't check `needsCustomLayoutForChildren` so it wrongly positions native views. You can see what I mean in the videos below. In the first video you can see the native close button is visible on the old architecture.  In the second video the close button is missing because the new React Native architecture has positioned the action view in the wrong place.

## Reproduction of the problem
I've created a [repository with a reproduction of the bug](https://github.com/grahammendick/needs-custom-layout-for-children-bug). The README in the repo contains the steps needed to recreate the problem. There are two apps in the repo, one built with the old architecture and the other built with the new architecture. You can see that the bug only happens on the new architecture. The only dependencies in the apps are the 3 packages that make up my navigation library.

You can see the close button appears on the old architecture

https://user-images.githubusercontent.com/1761227/175789921-d0cf0f51-e1ba-486c-82a4-f9141478e446.mov

But the close button is missing on the new architecture because React Native has positioned the action view in the wrong place

https://user-images.githubusercontent.com/1761227/175789924-23daeb08-f0d4-4823-982b-430674157d96.mov

**Comment by grahammendick** [comment, marked as answer]
Hi @sammy-SC and @cipolleschi I’ve been working on migrating my navigation library to the new architecture for about 6 months now and this bug is the final blocker! So I’ve freed up some time to work on the PR. But I’m going to need some guidance.

There’s a [TODO in the SurfaceMountingManager](https://github.com/facebook/react-native/blob/087f142cb79803cf0fdbe89e172176767b650b6e/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/SurfaceMountingManager.java#L995) that says
> TODO Check if the parent of the view has to layout the view, or the child has to lay itself out. see NativeViewHierarchyManager.updateLayout

So the call to `needsCustomLayoutForChildren` is missing from the `SurfaceMountingManager`. I've checked that adding this call would fix the bug shown in the videos above! But I don’t know how to add this call in for real because I can’t access the parent `ViewManager` without the `parentTag`. The `parentTag` isn't available in the new architecture like [it is in the old architecture](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/ReactAndroid/src/main/java/com/facebook/react/uimanager/NativeViewHierarchyManager.java#L210)

**Reply by grahammendick** [reply]
Hi @sammy-SC and @cipolleschi I've created a PR that is [my first attempt at migrating `needsCustomLayoutForChildren` to the new architecture ](https://github.com/facebook/react-native/pull/34254). It's not working but I'm hoping you can help me fix it.

The PR passes the parent tag to the `updateLayout` method of the `SurfaceMountingManager`. This works the first time the screen loads but the parent tag is 0 after I click the 'Hello' button (I'm using the repro fron this discussion to test it).

**Reply by grahammendick** [reply]
Hi @sammy-SC and @cipolleschi I've made a bit [more progress with the PR](https://github.com/facebook/react-native/pull/34254). The PR now passes the parent tag both first time and after I click the 'Hello' button. Here's the video of the bug fix running in the new architecture. You can see the action view is now positioned correctly so that the close button is visible.

The PR still needs work because I couldn't work out where to get the parent shadow view from in most cases. Please can you help me to get the PR into a mergeable state?

https://user-images.githubusercontent.com/1761227/180607896-35bf477f-4552-4b8a-8e09-9e8c49122c0c.mov

**Reply by cipolleschi** [reply]
Hi @grahammendick, thanks for working on this.
Unfortunately, I don't have a lot of context and knowledge about how these things work on Android.
@cortinico, could you help here, even by just tagging the proper person?

**Reply by grahammendick** [reply]
It's Ok it's already been merged in! Thanks anyway though 👍
